drake/solvers/Optimization.h

```diff
@@ -246,20 +278,27 @@ namespace Drake {
 //    const DecisionVariable& addIntegerVariables(size_t num_new_vars, std::string name);
 //  ...
 
-    void addCost(const std::shared_ptr<Constraint>& obj) {
-      checkVariables(obj);
+    void addCost(std::shared_ptr<Constraint> const& obj, VariableList const& vars) {
       problem_type.reset(problem_type->addGenericObjective());
-      generic_objectives.push_back(obj);
+      generic_objectives.push_back(Binding<Constraint>(obj, vars));
     }
 
-    std::shared_ptr<DifferentiableFunctionConstraint> addCost(const std::shared_ptr<DifferentiableFunction>& obj, const VariableList& vars) {
-      std::shared_ptr<DifferentiableFunctionConstraint> objective(new DifferentiableFunctionConstraint(vars,obj,1));
-      addCost(objective);
-      return objective;
+    void addCost(std::shared_ptr<Constraint> const& obj) {
+      addCost(obj, variable_views);
+    }
+
+    template <typename F>
+    typename std::enable_if<!std::is_convertible<F, std::shared_ptr<Constraint>>::value, std::shared_ptr<Constraint>>::type
+    addCost(F&& f, VariableList const& vars) {
+      auto c = std::make_shared<ConstraintImpl<F>>(std::forward<F>(f));
```

Comment:
While looking at support for passing function structures with varied ownership semantics I discovered that this needs to be revised further to distinguish lvalue and rvalue arguments better.  Please do not merge yet.